pkg/controller/direct/networkconnectivity/serviceconnectionpolicy_controller.go

```diff
@@ -0,0 +1,304 @@
+	networkConnectivityClient, err := m.newNetworkConnectivityClient(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO: Just fetch this object?
+	obj := &krm.NetworkConnectivityServiceConnectionPolicy{}
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(u.Object, &obj); err != nil {
+		return nil, fmt.Errorf("error converting to %T: %w", obj, err)
+	}
+
+	resourceID := ValueOf(obj.Spec.ResourceID)
+	if resourceID == "" {
+		resourceID = obj.GetName()
+	}
+	if resourceID == "" {
+		return nil, fmt.Errorf("cannot resolve resource ID")
+	}
+
+	location := obj.Spec.Location
+	if location == "" {
+		return nil, fmt.Errorf("cannot resolve location")
+	}
+
+	projectID := obj.Spec.ProjectRef.External
+	if projectID == "" {
+		return nil, fmt.Errorf("cannot resolve project")
+	}
+	{
+		tokens := strings.Split(projectID, "/")
+		if len(tokens) == 1 {
+			projectID = tokens[0]
+		} else if len(tokens) == 2 && tokens[0] == "projects" {
+			projectID = tokens[1]
+		} else {
+			return nil, fmt.Errorf("cannot resolve project from name %q", projectID)
+		}
+	}
+
+	mapCtx := &MapContext{
+		//	kube: kube,
+	}
+	desiredProto := ServiceConnectionPolicySpec_ToProto(mapCtx, &obj.Spec)
+	if mapCtx.Err() != nil {
+		return nil, mapCtx.Err()
+	}
+
+	return &serviceConnectionPolicyAdapter{
+		projectID:  projectID,
+		location:   location,
+		resourceID: resourceID,
+		// desired:          obj,
+		desiredProto: desiredProto,
+		gcpClient:    m.gcpClient,
+		client:       networkConnectivityClient,
+	}, nil
+}
+
+// Find implements the Adapter interface.
+func (a *serviceConnectionPolicyAdapter) Find(ctx context.Context) (bool, error) {
+	if a.resourceID == "" {
+		return false, nil
+	}
+
+	name := a.fullyQualifiedName()
+	found, err := a.client.Projects.Locations.ServiceConnectionPolicies.Get(name).Context(ctx).Do()
+	if err != nil {
+		if IsNotFound(err) {
+			return false, nil
+		}
+		return false, err
+	}
+
+	a.actual = found
+
+	return true, nil
+}
+
+// Delete implements the Adapter interface.
+func (a *serviceConnectionPolicyAdapter) Delete(ctx context.Context) (bool, error) {
+	// Already deleted
+	if a.resourceID == "" {
+		return false, nil
+	}
+
+	// TODO: Delete via status selfLink?
+
+	op, err := a.client.Projects.Locations.ServiceConnectionPolicies.Delete(a.fullyQualifiedName()).Context(ctx).Do()
+	if err != nil {
+		if IsNotFound(err) {
+			return false, nil
+		}
+		return false, fmt.Errorf("deleting serviceConnectionPolicy %s: %w", a.fullyQualifiedName(), err)
+	}
+
+	if err := a.waitForOp(ctx, op); err != nil {
+		return false, fmt.Errorf("waiting for serviceConnectionPolicy delete %s: %w", a.fullyQualifiedName(), err)
+	}
+
+	return true, nil
+}
+
+func (a *serviceConnectionPolicyAdapter) waitForOp(ctx context.Context, op *api.GoogleLongrunningOperation) error {
+	// TODO: Only wait a short time, then update status?
+	for {
```

Comment:
Can this wait forever?

Reply:
Yes.  We need to figure out the right behaviour here (e.g. ideally we would update status during the update, ideally we would not block one of our reconciler threads etc)